keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+  - Each IPv4 service IP will be mapped to all IPv4 addresses on backend pods
+  - Each IPv6 service IP will be mapped to all IPv6 addresses on backend pods
+
+(Note that this assumed behavior works very well for the typical dual-stack case where there are one IPv4 and one IPv6 service CIDR, and one IPv4 and one IPv6 address per pod.)
+
+#### Versioned API Change: Service v1 core
+In order to maintain backwards compatibility for the core V1 API, this proposal adds a new "ExtraClusterIPs" field to the core V1 version of the [Service V1 core API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#service-v1-core), while maintaining the existing "ClusterIP" field. This change is in accordance with the [Kubernetes API change
+quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md):
+
+```
+    // clusterIP is the IP address of the service and is usually assigned
+    // randomly by the master.
+    ClusterIP string `json:"clusterIP,omitempty" protobuf:"bytes,3,opt,name=clusterIP"`
+
+    // Additional cluster IPs
+    ExtraClusterIPs []string `json:"extraclusterIP,omitempty" protobuf:"bytes,3,opt,name=ExtraclusterIPs"`
+```
+
+#### Internal Representation:
+The internal representation for the service IP (a.k.a. cluster IP) will be changed from a singular string to a slice of strings, and renamed "ClusterIPs":
+```
+    ClusterIPs []string `json:"clusterIPs,omitempty" protobuf:"bytes,3,opt,name=clusterIPs"`
+```
+
+#### Changes to the Service IP Allocator
+- The service allocator will need to have one or more prefixes specified for it to allocate service IPs
+- The service allocator will allocate and track service IPs for each prefix
+
+#### kube-apiserver Startup Configuration for Multiple Service CIDRs
+A new, plural "service-cluster-ip-ranges" option for the [kube-apiserver startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/) is proposed, in addition to retaining the existing, singular "service-cluster-ip-range" option (for backwards compaibility):
+```
+  --service-cluster-ip-range  ipNet         [Singular IP CIDR,  Default: 10.0.0.0/24]
+  --service-cluster-ip-ranges stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --service-cluster-ip-range argument will become deprecated.
+
+#### controller-manager Startup Configuration for Multiple Service CIDRs
+A new, plural "service-cluster-ip-ranges" option for the [controller-manager startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/) is proposed, in addition to retaining the existing, singular "service-cluster-ip-range" option (for backwards compaibility):
+```
+  --service-cluster-ip-range  ipNet         [Singular IP CIDR,  Default: 10.0.0.0/24]
+  --service-cluster-ip-ranges stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
```

Comment:
what happens if the user specifies two v4 ranges?

Reply:
good call - err on the side of over-specifying :)